### PR TITLE
[SPARK-12860][SQL] speed up safe projection for primitive types

### PR DESCRIPTION
**Pull request by cloud-fan**
The idea is simple, use `SpecificMutableRow` instead of `GenericMutableRow` as result row for safe projection.

A simple benchmark shows about 1.5x speed up for primitive types, code: https://gist.github.com/cloud-fan/fa77713ccebf0823b2ab#file-safeprojectionbenchmark-scala

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @davies

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -495,6 +495,7 @@ abstract class GeneratedClass {
 abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Logging {
 
   protected val genericMutableRowType: String = classOf[GenericMutableRow].getName
+  protected val specificMutableRowType: String = classOf[SpecificMutableRow].getName
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: This is only used by SafeProjection, we could put it there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateSafeProjection.scala [hunk 2]
```diff
@@ -180,6 +182,10 @@ object GenerateSafeProjection extends CodeGenerator[Seq[Expression], Projection]
     logDebug(s"code for ${expressions.mkString(",")}:\n${CodeFormatter.format(code)}")
 
     val c = CodeGenerator.compile(code)
-    c.generate(ctx.references.toArray).asInstanceOf[Projection]
+    c.generate(ctx.references.toArray ++ expressions.map(_.dataType)).asInstanceOf[Projection]
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateSafeProjection.scala]: We could pass a SpecificMutableRow into it

**Comment by davies** [comment]
merging this into master, thanks!

**Pull request closed without merging**